<?xml version="1.0"?>
<robot name="robot_arm">

<!-- only movable links have number in the name, like link1, link2 -->
<!-- only movable links have number in the name, like joint1, joint2 -->

<!-- world link -->
<link name="world"/>

<!-- base_link and its fixed joint -->
<joint name="joint_fix" type="fixed">
	<parent link="world"/>
	<child link="base_link"/>
</joint>

<link name="base_link">
	<collision>
		<origin xyz="0 0 0.75" rpy="0 0 0"/>
		<geometry>
			<box size="0.2 0.2 1.5"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0.75" rpy="0 0 0"/>
		<geometry>
			<box size="0.2 0.2 1.8"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0.75" rpy="0 0 0"/>
		<mass value="1"/>
		<inertia
			ixx="1.0" ixy="0.0" ixz="0.0"
			iyy="1.0" iyz="0.0"
			izz="1.0"/>
	</inertial>
</link>

<!-- link1 and its revolute joint1 relative to base_link -->
<!-- rotating along z axis -->
<joint name="joint1" type="revolute">
	<parent link="base_link"/>
	<child link="link1"/>
	<origin xyz="0 0 1.5" rpy="0 0 0"/>
	<axis xyz="0 0 1"/>
	<limit effort="100" velocity="100" lower="-1.5708" upper="1.5708"/>
</joint>

<link name="link1">
	<collision>
		<origin xyz="0 0 0.05" rpy="0 0 0"/>
		<geometry>
			<cylinder length="0.1" radius="0.1"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0.05" rpy="0 0 0"/>
		<geometry>
			<cylinder length="0.1" radius="0.1"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0.05" rpy="0 0 0"/>
		<mass value="1"/>
		<inertia
			ixx="0.001" ixy="0.0" ixz="0.0"
			iyy="0.001" iyz="0.0"
			izz="0.001"/>
	</inertial>
</link>

<gazebo reference="link1">
	<material>Gazebo/Grey</material>
</gazebo>

<!-- link2 and its revolute joint2 relative to link1 -->
<!-- rotating along y axis -->
<joint name="joint2" type="revolute">
	<parent link="link1"/>
	<child link="link2"/>
	<origin xyz="0 0 0.1" rpy="0 0 0"/>
	<axis xyz="0 1 0"/>
	<limit effort="100" velocity="100" lower="-0.1" upper="3.1416"/>
</joint>

<link name="link2">
	<collision>
		<origin xyz="0 0 0.6" rpy="0 0 0"/>
		<geometry>
			<cylinder length="1.1" radius="0.09"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0.6" rpy="0 0 0"/>
		<geometry>
			<cylinder length="1.2" radius="0.09"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0.6" rpy="0 0 0"/>
		<mass value="1"/>
		<inertia
			ixx="0.1" ixy="0.0" ixz="0.0"
			iyy="0.1" iyz="0.0"
			izz="0.005"/>
	</inertial>
</link>

<gazebo reference="link2">
	<material>Gazebo/Grey</material>
</gazebo>

<!-- link3 and its revolute joint3 relative to link2 -->
<!-- rotating along y axis -->
<joint name="joint3" type="revolute">
	<parent link="link2"/>
	<child link="link3"/>
	<origin xyz="0 0 1.2" rpy="0 0 0"/>
	<axis xyz="0 1 0"/>
	<limit effort="100" velocity="100" lower="-1.5708" upper="1.5708"/>
</joint>

<link name="link3">
	<collision>
		<origin xyz="0 0 0.15" rpy="0 0 0"/>
		<geometry>
			<cylinder length="0.3" radius="0.07"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0.15" rpy="0 0 0"/>
		<geometry>
			<cylinder length="0.3" radius="0.07"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0.15" rpy="0 0 0"/>
		<mass value="1"/>
		<inertia
			ixx="0.1" ixy="0.0" ixz="0.0"
			iyy="0.1" iyz="0.0"
			izz="0.005"/>
	</inertial>
</link>

<gazebo reference="link3">
	<material>Gazebo/Grey</material>
</gazebo>

<!-- link4 and its revolute joint4 relative to link3 -->
<!-- rotating along y axis -->
<joint name="joint4" type="revolute">
	<parent link="link3"/>
	<child link="link4"/>
	<origin xyz="0 0 0.3" rpy="0 0 0"/>
	<axis xyz="1 0 0"/>
	<limit effort="100" velocity="100" lower="-1.5708" upper="1.5708"/>
</joint>

<link name="link4">
	<collision>
		<origin xyz="0 0 0.06" rpy="0 0 0"/>
		<geometry>
			<cylinder length="0.12" radius="0.05"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0.06" rpy="0 0 0"/>
		<geometry>
			<cylinder length="0.12" radius="0.05"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0.06" rpy="0 0 0"/>
		<mass value="0.4"/>
		<inertia
			ixx="0.05" ixy="0.0" ixz="0.0"
			iyy="0.002" iyz="0.0"
			izz="0.05"/>
	</inertial>
</link>

<gazebo reference="link4">
	<material>Gazebo/Red</material>
</gazebo>


<!-- base_link and its fixed joint -->
<joint name="joint_fix_p" type="fixed">
  <parent link="world"/>
  <child link="base_link_p"/>
</joint>

<link name="base_link_p">
  <collision>
    <origin xyz="0 1.5 0.05" rpy="0 0 0"/>
    <geometry>
      <box size="1 1 0.1"/>
    </geometry>
  </collision>

  <visual>
    <origin xyz="0 1.5 0.05" rpy="0 0 0"/>
    <geometry>
      <box size="1 1 0.1"/>
    </geometry>
  </visual>

  <inertial>
    <origin xyz="0 1.5 0.5" rpy="0 0 0"/>
    <mass value="1"/>
    <inertia
      ixx="1.0" ixy="0.0" ixz="0.0"
      iyy="1.0" iyz="0.0"
      izz="1.0"/>
  </inertial>
</link>

<!-- link1 and its revolute joint1 relative to base_link -->
<!-- rotating along z axis -->
<joint name="joint5" type="revolute">
  <parent link="base_link_p"/>
  <child link="link1_p"/>
  <origin xyz="0 1.5 0.15" rpy="0 0 0"/>
  <axis xyz="0 0 1"/>
  <limit effort="100" velocity="100" lower="-3.1416" upper="3.1416"/>
</joint>

<link name="link1_p">
  <collision>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <box size="1 1 0.1"/>
    </geometry>
  </collision>

  <visual>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <box size="1 1 0.1"/>
    </geometry>
  </visual>

  <inertial>
    <origin xyz="0 0 0.05" rpy="0 0 0"/>
    <mass value="1"/>
    <inertia
      ixx="1" ixy="0.0" ixz="0.0"
      iyy="1" iyz="0.0"
      izz="1"/>
  </inertial>
</link>

<gazebo reference="link1_p">
  <material>Gazebo/Grey</material>
</gazebo>

<joint name="joint_box" type="fixed">
  <parent link="link1_p"/>
  <child link="box"/>
</joint>

<link name="box">
  <collision>
    <origin xyz="0 0 0.3" rpy="0 0 0"/>
    <geometry>
      <box size="0.5 0.5 0.5"/>
    </geometry>
  </collision>

  <visual>
    <origin xyz="0 0 0.3" rpy="0 0 0"/>
    <geometry>
      <box size="0.5 0.5 0.5"/>
    </geometry>
  </visual>

  <inertial>
    <origin xyz="0 0 0.05" rpy="0 0 0"/>
    <mass value="1"/>
    <inertia
      ixx="1" ixy="0.0" ixz="0.0"
      iyy="1" iyz="0.0"
      izz="1"/>
  </inertial>
</link>

<gazebo reference="box">
  <material>Gazebo/black</material>
</gazebo>


</robot>

